### ISSUE TITLE
iaf_psc_alpha_ps model stuck when simulating

### ISSUE THREAD

**Issue by stinebuu** [nest/nest-simulator#1272]
When running the `brunel_ps.sli` example with current master (after having changed `tau_syn` to `tau_syn_ex` and `tau_syn_in`), the example never finish simulating. The problem is the `iaf_psc_alpha_ps` model, which seems to be stuck in the loop in the `bisectioning_` function. If I try to use the `iaf_psc_alpha_canon` model, the example runs through without problems.

For a simpler test, one can also look at `test_mini_brunel_ps.sli`, which is currently using the `iaf_psc_alpha_canon` model. But if we instead use `iaf_psc_alpha_ps`, we again get stuck when simulating. I have tried to change the threshold in the `bisection_` function to `1e-6`, and this does not help. If I try to print a counter in the while loop, the counter keeps on increasing and it seems likely that this might be the problem and that we are not able to leave the loop.

**Comment by heplesser** [comment]
See also #403.

**Comment by heplesser** [comment]
@stinebuu Since this issue has surfaced again in #1440, could you take another look and in addition to a counter also follow how `root` and `V_m_root` develop? As `div` grows, at some point `dt / div` will be so small that adding/subtracting to/from `root` will no longer change `root`,  and then we are definitely stuck. Before we make any changes, we should understand a little better under which circumstances these problems occur. One solution would be a softer termination condition, the other termination after a given number of iterations.

**Comment by stinebuu** [comment]
@heplesser I am not really able to find a pattern in when things go wrong. I have run a simplified version of `test_brunel_mini_ps.sli` with `iaf_psc_alpha_ps` instead of `iaf_psc_alpha_canon`, and looked at when `bisection_` converges and when it does not. I have tried using all default parameter values for `iaf_psc_alpha_ps`, and also tried using specific `tau_m`, `V_m` etc., and both versions gets stuck every time. 

The following two examples have very similar `V_m_root` and `fabs( P_.U_th_ - V_m_root )` at first iteration, but while the first example converges after 36 iterations, the second has not converged after 1000 iterations, at which point I have tell `bisection_` to continue. `t_step` below is `dt / div^{iter}`

Converges:
```
counter: 0 dt:  0.0681367 root: 0.0340683 t_step 0.0340683   P_.U_th_: 15 V_m_root: 14.9977 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 0.0023093
counter: 1 dt:  0.0681367 root: 0.0511025 t_step 0.0170342   P_.U_th_: 15 V_m_root: 15.0008 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 0.000847078
counter: 2 dt:  0.0681367 root: 0.0425854 t_step 0.00851709  P_.U_th_: 15 V_m_root: 14.9993 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 0.000699876
counter: 3 dt:  0.0681367 root: 0.046844  t_step 0.00425854  P_.U_th_: 15 V_m_root: 15.0001 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 8.139e-05
counter: 4 dt:  0.0681367 root: 0.0447147 t_step 0.00212927  P_.U_th_: 15 V_m_root: 14.9997 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 0.000307293
counter: 5 dt:  0.0681367 root: 0.0457793 t_step 0.00106464  P_.U_th_: 15 V_m_root: 14.9999 V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 0.000112465
counter: 6 dt:  0.0681367 root: 0.0463117 t_step 0.000532318 P_.U_th_: 15 V_m_root: 15      V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 1.54156e-05
counter: 7 dt:  0.0681367 root: 0.0465778 t_step 0.000266159 P_.U_th_: 15 V_m_root: 15      V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 3.30176e-05
counter: 8 dt:  0.0681367 root: 0.0464447 t_step 0.000133079 P_.U_th_: 15 V_m_root: 15      V_.V_m_before_: 14.9906 fabs( P_.U_th_ - V_m_root ) 8.80864e-06
```

Does not converge:
```
counter: 0  dt: 0.1 root: 0.05      t_step 0.05        P_.U_th_: 15 V_m_root: 14.9978 V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00224008
counter: 1  dt: 0.1 root: 0.075     t_step 0.025       P_.U_th_: 15 V_m_root: 14.9981 V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.001914
counter: 2  dt: 0.1 root: 0.0875    t_step 0.0125      P_.U_th_: 15 V_m_root: 14.9981 V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00192767
counter: 3  dt: 0.1 root: 0.09375   t_step 0.00625     P_.U_th_: 15 V_m_root: 14.998  V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00197829
counter: 4  dt: 0.1 root: 0.096875  t_step 0.003125    P_.U_th_: 15 V_m_root: 14.998  V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.0020145
counter: 5  dt: 0.1 root: 0.0984375 t_step 0.0015625   P_.U_th_: 15 V_m_root: 14.998  V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00203533
counter: 6  dt: 0.1 root: 0.0992188 t_step 0.00078125  P_.U_th_: 15 V_m_root: 14.998  V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00204642
counter: 7  dt: 0.1 root: 0.0996094 t_step 0.000390625 P_.U_th_: 15 V_m_root: 14.9979 V_.V_m_before_: 14.9957 fabs( P_.U_th_ - V_m_root ) 0.00205214
```

It does not seem to have anything to do with `dt`, (which `root` is dependent of), because while `dt=0.1` does not converge above, this verision of `dt=0.1` *does* converge, after 37 iterations:
```
counter: 0  dt: 0.1 root: 0.05      t_step 0.05        P_.U_th_: 15 V_m_root: 14.9999 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 6.3995e-05
counter: 1  dt: 0.1 root: 0.075     t_step 0.025       P_.U_th_: 15 V_m_root: 15.0035 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 0.00353709
counter: 2  dt: 0.1 root: 0.0625    t_step 0.0125      P_.U_th_: 15 V_m_root: 15.0018 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 0.0018019
counter: 3  dt: 0.1 root: 0.05625   t_step 0.00625     P_.U_th_: 15 V_m_root: 15.0009 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 0.000885351
counter: 4  dt: 0.1 root: 0.053125  t_step 0.003125    P_.U_th_: 15 V_m_root: 15.0004 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 0.000414785
counter: 5  dt: 0.1 root: 0.0515625 t_step 0.0015625   P_.U_th_: 15 V_m_root: 15.0002 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 0.000176423
counter: 6  dt: 0.1 root: 0.0507813 t_step 0.00078125  P_.U_th_: 15 V_m_root: 15.0001 V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 5.6471e-05
counter: 7  dt: 0.1 root: 0.0503906 t_step 0.000390625 P_.U_th_: 15 V_m_root: 15      V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 3.69775e-06
counter: 8  dt: 0.1 root: 0.0505859 t_step 0.000195313 P_.U_th_: 15 V_m_root: 15      V_.V_m_before_: 14.9911 fabs( P_.U_th_ - V_m_root ) 2.64027e-05
```

The only thing I see is that I think maybe `root` needs to stay relatively close to `root_0`. The converging example right above has `root_0 = 0.5` and `root_37 = 0.0504146...` (converges for iter = 37) and the non-convergent example has `root_0 = 0.5` and `root_1000 = 0.1`. Another converging instance has `root_0 = 0.0423023` and `root_38 = 0.0642229...`, so a bit more change, but not super much. The first converging example listed above has `root_0 = 0.0340683`, `root_36 = 0.0463963`.

Another thing, for the example I am running now, if `bisection_` has not converged after around 40 iterations, it will not converge at all. This will vary from example to example of course, but it seems like the number of iterations for convergence within a given example is pretty consistent.

I don't really know what to look for now, do you see anything or have any tips?

I can always add termination after a given number of iterations, I think a lower tolerance is a problem because we have to lower it a lot. The tolerance would need to be around `1e-2` while it is `1e-14` now.

**Comment by heplesser** [comment]
@stinebuu Interesting. Three observations:
1. After the first loop iteration, we will have `root == dt/2`. We then add/subtract terms that shrink by powers of 2, so after 52 rounds at the latest, what we add or subtract will be less than machine precision and lost, and `root` remains constant for all further iterations.
1. In the second case above (non-converging), the error increases again from step 2 onwards, and ins step 7 `V_m_root` is actually going down again (away from the threshold). This is rather strange and needs closer investigation.

One thing I am wondering about is what would happen if we enter `bisectioning_()` in a situation where `S.V_m_ == P.U_th_` or really only minimally larger, i.e., a situation where the threshold crossing is at the very end of the interval. That case we would handle badly, because we start with `root == 0`, i.e., at the left edge of the interval and then move right in increasingly smaller steps. We also iterate in a somewhat "dumb" fashion since our step does not depend on function values. I think using [*regula falsi* or a variant](https://en.wikipedia.org/wiki/Regula_falsi) might work better.

Another issue is that in computing the new `V_m_root`, we have a term `expm1_tau_m * V_.V_m_before_ + V_.V_m_before_`, which would better be expressed as `V_.V_m_before * exp(-root/P_.tau_m_)` to avoid adding two terms of similar size. The same hold for all `V_m_`-update equations in the model.

Finally, only in `bisectioning_()` do we use `propagator_32()` to deal with possible singularities. We should do this in all other places updating `V_m_` as well.

**Comment by stinebuu** [comment]
> Another issue is that in computing the new V_m_root, we have a term expm1_tau_m * V_.V_m_before_ + V_.V_m_before_, which would better be expressed as V_.V_m_before * exp(-root/P_.tau_m_) to avoid adding two terms of similar size. The same hold for all V_m_-update equations in the model.

@heplesser So we should no longer add the second `V_.V_m_before` term? Or should I just replace `expm1_tau_m` with `exp(-root/P_.tau_m_)`?

I'll look into regula falsi and your other comments now.

**Comment by heplesser** [comment]
Status update: @stinebuu and I just analysed the code and understand the problem. This is a real bug: bisectioning is based on exp- instead of alpha-dynamics in `iaf_psc_alpha_ps`. A fix is on its way and should be included in #1455.
